extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriGatewayHandler.java

```diff
@@ -105,16 +138,45 @@ public void initialize() {
         }
 
         DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder(new InetSocketAddress(0));
-        builder.setPskStore(new StaticPskStore("", configuration.code.getBytes()));
+        builder.setPskStore(new StaticPskStore(configuration.identity, configuration.preSharedKey.getBytes()));
         dtlsConnector = new DTLSConnector(builder.build());
         endPoint = new CoapEndpoint(dtlsConnector, NetworkConfig.getStandard());
         deviceClient.setEndpoint(endPoint);
         updateStatus(ThingStatus.UNKNOWN);
 
         // schedule a new scan every minute
-        scanJob = scheduler.scheduleWithFixedDelay(() -> {
-            startScan();
-        }, 0, 1, TimeUnit.MINUTES);
+        scanJob = scheduler.scheduleWithFixedDelay(this::startScan, 0, 1, TimeUnit.MINUTES);
+    }
+
+    protected String authenticateOnGateway(TradfriGatewayConfig configuration) {
+        String preSharedPsk = null;
+
+        try {
+            DtlsConnectorConfig.Builder builder = new DtlsConnectorConfig.Builder(new InetSocketAddress(0));
+            builder.setPskStore(new StaticPskStore("Client_identity", configuration.code.getBytes()));
+
+            DTLSConnector dtlsConnector = new DTLSConnector(builder.build());
+            CoapEndpoint authEndpoint = new CoapEndpoint(dtlsConnector, NetworkConfig.getStandard());
+            String authUrl = "coaps://" + configuration.host + ":" + configuration.port + "/15011/9063";
+
+            CoapClient deviceClient = new CoapClient(new URI(authUrl));
+            deviceClient.setEndpoint(authEndpoint);
+
+            JsonObject json = new JsonObject();
+            json.addProperty(CLIENT_IDENTITY_PROPOSED, configuration.identity);
+
+            CoapResponse r = deviceClient.post(json.toString(), 0);
+            String responseText = r.getResponseText();
+            authEndpoint.destroy();
+            deviceClient.shutdown();
+
+            json = new JsonParser().parse(responseText).getAsJsonObject();
+            preSharedPsk = json.get(NEW_PSK_BY_GW).getAsString();
+        } catch (URISyntaxException | JsonParseException e) {
+            logger.error("Can't obtain pre-shared key", e);
```

Comment:
Note that errors must only be logged if we run into a bug situation in your code - so the expectation is that we should never get here.
If so, I am missing all error handling about the expected cases like communication timeouts, gateways that don't accept the secret code or respond with an CoAP error, etc.
You will need to set different Thing statuses for those different situations.

Reply:
Looking at the current version of the code - all those:
> all error handling about the expected cases like communication timeouts, gateways that don't accept the secret code or respond with an CoAP error, etc.

Are handled in only one place - the ```onError``` callback of the ```TradfriCoapHandler``` and this hasn't changed so the end user won't notice any regression.

Reply:
We are within the `authenticateOnGateway` method here, for which there is no "current version" - this is a new feature. Apart from this, I cannot see `TradfriCoapHandler` being used here at all.